core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -1951,6 +1952,20 @@ public static TimeZone timeZone(DataContext root) {
     return (TimeZone) DataContext.Variable.TIME_ZONE.get(root);
   }
 
+  /** SQL {@code USER} function. */
+  @NonDeterministic
+  public static String user(DataContext root) {
+    return Optional.ofNullable((String) DataContext.Variable.USER.get(root))
+        .orElse("sa");
+  }
+
+  /** SQL {@code SYSTEM_USER} function. */
+  @NonDeterministic
```

Comment:
Does it really need `NonDeterministic` here?
I guess the function is more-or-less pure as it just reads the value of data context.

Reply:
This is stay consistent with the other dynamic functions. I have no idea why any of them are annotated this way but felt the new ones should be consistent.

Reply:
Conceivably (I know it's a stretch) a query could be prepared by one user and executed by another. The NonDeterministic prevents constant-reduction.